draft-pda-core.md

```diff
@@ -291,11 +293,439 @@ Each PA task in this document is divided into three sub-protocols as follows.
+to ensure that the helper can reproduce this transcript. This should be possible
+given the information it has.]
+
+Next, the client encodes its measurements as an input for the PA protocol. It
+then generates a validity proof for its input and uses `context` to split the
+input and proof into a *leader share* and a *helper share*, where the latter is
+protected by the HPKE context. Note that the details of each of these processing
+steps --- encode, prove, split, and encrypt --- are specific to the PA protocol.
+
+[OPEN ISSUE: Is it safe to generate the proof once, then secret-share between
+each (leader, helper) pair? Probably not in general, but maybe for Prio?]
+
+The payload of the POST request to `[leader]/upload_finish` is structured as
+follows:
+
+~~~
+struct {
+  PATask task;                 // Equal to PAUploadStartReq.task
+  opaque id[16];               // Equal to PAUploadStartReq.id
+  uint8 helper_hpke_config_id; // Equal to HpkeConfig.id
+  Url helper_url;
+  PAHelperShare helper_share;
+  PALeaderShare leader_share;
+} PAUploadFinishReq;
+~~~
+
+We sometimes refer to this message as the *report*. The message contains the
+`task` and `id` fields of the previous request. In addition, it includes the
+helper's HPKE config id, endpoint URL, and the helper and leader shares.
+The helper share has the following structure:
+
+~~~
+struct {
+  opaque enc<1..2^16-1>; // Equal to helper_enc
+  PAProto proto;
+  select (PAHelperShare.proto) {
+    case prio: PrioHelperShare;
+    case hits: HitsHelperShare;
+  }
+} PAHelperShare;
+~~~
+
+Field `enc` encodes the helper's encapsulated HPKE context. The remainder of the
+structure contains the share itself, the structure of which is specific to the
+PA protocol. The structure of the leader share is similarly protocol specific:
+
+~~~
+struct {
+  PAProto proto;
+  select (PALeaderShare.proto) {
+    case prio: PrioLeaderShare;
+    case hits: HitsLeaderShare;
+  };
+} PALeaderShare;
+~~~
+
+Note that the leader share is sent not encrypted.
+
+**Response.**
+The leader responds to well-formed requests to `[leader]/upload_finish` with
+status 200 and an empty payload. Malformed requests are handled as described in
+{{pa-error-common-aborts}}.
+
+## Verify {#pa-verify}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+After the client uploads a report to the leader, the leader and helper verify in
+zero knowledge that the proof is well-formed. The exact procedure for doing so
+is protocol specific, but all protocols have the same basic structure. In
+particular, the protocol is comprised of a sequence of HTTPS requests from the
+leader to the helper. At the end of this phase, the leader and helper will have
+decided whether a set of client inputs are valid. For each valid input, they
+proceed as described in {{pa-collect}}.
+
+**Request.**
+The leader begins by collecting a sequence of reports that are all associated
```

Comment:
```suggestion
The leader begins with a sequence of reports that are all associated
```

Slightly reworded so as to not require servers to do some sort of batching if they don't have any reports to batch.